site/_posts/2017-09-21-finalizers-can-be-interrupted-from-time-to-time.md

```diff
@@ -0,0 +1,179 @@
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/httpclient-2.8.3/lib/httpclient/session.rb:      ::Timeout.timeout(@receive_timeout, ReceiveTimeoutError) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/httpclient-2.8.3/lib/httpclient/session.rb:        ::Timeout.timeout(@receive_timeout, ReceiveTimeoutError) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/httpclient-2.8.3/lib/httpclient/session.rb:       ::Timeout.timeout(@receive_timeout, ReceiveTimeoutError) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/httpclient-2.8.3/lib/httpclient/session.rb:        ::Timeout.timeout(@receive_timeout, ReceiveTimeoutError) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/manageiq-network_discovery-0.1.2/lib/manageiq/network_discovery/port_scanner.rb:          Timeout.timeout(ost.timeout) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/manageiq-smartstate-0.1.4/lib/metadata/VmConfig/VmConfig.rb:      Timeout.timeout(60) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/multipart-post-2.0.0/test/test_composite_io.rb:      Timeout.timeout(1) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/net-ping-1.7.8/lib/net/ping/icmp.rb:        Timeout.timeout(@timeout){
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/net-ping-1.7.8/lib/net/ping/udp.rb:        Timeout.timeout(@timeout){
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/nio4r-2.1.0/spec/nio/selector_spec.rb:        Timeout.timeout(2) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/nio4r-2.1.0/spec/spec_helper.rb:      sock = Timeout.timeout(1) { TCPSocket.new("localhost", $current_tcp_port) }
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/open4-1.3.4/lib/open4.rb:    Timeout.timeout(timeout){ yield }
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/rack-2.0.3/test/spec_utils.rb:    Timeout.timeout(1) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/ruby-dbus-0.13.0/spec/spec_helper.rb:    Timeout.timeout(10) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/snmp-1.2.0/lib/snmp/manager.rb:          Timeout.timeout(@timeout) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/vcr-3.0.3/lib/vcr/util/internet_connection.rb:          Timeout.timeout(timeout) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/webmock-2.3.2/spec/support/network_connection.rb:    Timeout.timeout(timeout) do
+
+05:36:21 ~/Code/manageiq (master) (2.4.2) + grep -r -E "\bTimeout::timeout\b" /Users/joerafaniello/.gem/ruby/2.4.2/gems
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/fog-core-1.45.0/lib/fog/core/current_machine.rb:    #     Timeout::timeout(5) do
+/Users/joerafaniello/.gem/ruby/2.4.2/gems/open4-1.3.4/lib/open4.rb:          Timeout::timeout(timeout) do
+```
+
+Those are only gems used by ManageIQ.  I'm sure there are many other gems used by other
+projects that still use Timeout.
+
+Wait a second.  I didn't actually look at the contents of the `Timeout.timeout` block.
+That's the point, it doesn't matter.  If you hit that code at all, there's a small
+chance it can `Thread#raise` at a very inopportune time in code that will never
+expect it!
```

Comment:
While documented before, I had read 2 of the 3 above posts and still was surprised when I finally understood the problem of, so I'd spend a few words to explain.  Here looks a good place. E.g:

> You might be thinking "I'm expecting it, my critical code is wrapped by `begin..rescue/ensure` cleaning things up".  It's true, it's a regular exception and will be rescued if raised inside that critical code, but don't underestimate Murphy's law.  What if it's raised inside the *cleanup* code preventing it from completing?  And that's not the end of it, it will find the weirdest possible place (on Friday).